pkg/k8s/init.go

```diff
@@ -271,3 +280,101 @@ func RegisterCRDs() error {
 
 	return nil
 }
+
+// WaitForCRD waits for the given CRD to be available until the default timeout,
+// after which cilium-agent should be ready. Returns an error when timeout
+// is exceeded.
+func WaitForCRD(client clientset.Interface, name string) error {
+	ctx, cancelFunc := context.WithTimeout(context.Background(), option.Config.CRDWaitTimeout)
+	defer cancelFunc()
+	return waitForCRD(ctx, client, name)
+}
+
+// waitForCRD waits for the given CRD to be available with the given context.
+func waitForCRD(ctx context.Context, client clientset.Interface, name string) error {
+	log.Infof("Waiting for CRD %s to be available", name)
+
+	selector := fields.OneTermEqualSelector("metadata.name", name).String()
```

Comment:
instead of creating a single "waitForCRD" for each CRD we should watch for all CRDs. We should also make sure that the watcher does not download the entire CRD for all agents.

In curl it's easy to do it by passing this headers (specially the `as=Table`):
```
curl -k -v -XGET  -H "Accept: application/json;as=Table;v=v1;g=meta.k8s.io,application/json;as=Table;v=v1beta1;g=meta.k8s.io,application/json" -H "User-Agent: kubectl/v1.19.1 (linux/amd64) kubernetes/206bcad" 'https://192.168.33.11:6443/apis/apiextensions.k8s.io/v1/customresourcedefinitions?resourceVersion=1034518&watch=true'
```

otherwise all cilium agents will download the CRD which is unnecessary since we only care about its name

Reply:
Synced offline. The gist was that we needed to implement a custom controller of sorts in order to retrieve CRDs in an efficient manner.